package.json

```diff
@@ -2,7 +2,18 @@
   "name": "single-spa-react",
   "version": "2.14.0",
   "description": "A single spa plugin for React apps",
-  "main": "lib/single-spa-react.js",
+  "main": "lib/umd/single-spa-react.js",
+  "module": "lib/esm/single-spa-react.js",
+  "exports": {
+    "import": {
+      ".": "./lib/esm/single-spa-react.js",
+      "./parcel": "./lib/esm/parcel.js"
```

Comment:
The `exports` field is really cool, but also is rather tricky to get right. @frehner can you comment on this?

@madeofsun have you verified that NodeJS 14 can both `import` and `require` the main and parcel entries? You can do so I think with `npm link` in a test project.  Not sure if there's an easier way to do it.

Reply:
I does not work. Where can I read about it, because it seems hard to find documentation about this field.

Reply:
https://nodejs.org/api/esm.html#esm_package_entry_points

Reply:
ugh yeah I'm still not very clear on what should be used and how. 

for example, do we also want/need to add a `node: {}` and `browser: {}`? it's unclear, but it doesn't appear that we need to - but could.

Reply:
The problem was that the path directive should be before the conditional directive:

```json
"exports": {
    ".": {
      "import": "./lib/esm/single-spa-react.js",
      "require": "./lib/umd/single-spa-react.js"
    },
    "./parcel": {
      "import": "./lib/esm/parcel.js",
      "require": "./lib/umd/parcel.js"
    }
  }
```

But I have faced another problem. There is a naming convention with `cjs`, `mjs` and `js`. 
In order to use it as ESM, the file names for esm entries should end with `.mjs` or the package type should be "module", but then CommonJS entries should end with `.cjs`:
```json
"exports": {
    ".": {
      "import": "./lib/esm/single-spa-react.mjs",
      "require": "./lib/umd/single-spa-react.mjs"
    },
    "./parcel": {
      "import": "./lib/esm/parcel.js",
      "require": "./lib/umd/parcel.js"
    }
  }
```
OR
```json
"type": "module",
"exports": {
    ".": {
      "import": "./lib/esm/single-spa-react.js",
      "require": "./lib/umd/single-spa-react.js"
    },
    "./parcel": {
      "import": "./lib/esm/parcel.cjs",
      "require": "./lib/umd/parcel.cjs"
    }
  }
```
OR should we do like this:
```json
"exports": {
    ".": {
      "import": "./lib/esm/single-spa-react.mjs",
      "require": "./lib/umd/single-spa-react.cjs"
    },
    "./parcel": {
      "import": "./lib/esm/parcel.mjs",
      "require": "./lib/umd/parcel.cjs"
    }
  }
```

Then there is another question arises. Should we have copies of the same file with different names, f.e: `./lib/esm/single-spa-react.mjs` and `./lib/esm/single-spa-react.js`

Reply:
> do we also want/need to add a node: {} and browser: {}

Up until now, single-spa-react hasn't actively supported being executed in NodeJS, but I think that it should. With server rendering, it would be desireable. However, for now I don't think we need to differentiate between node/browser in the package.json exports, since we don't have any code different between the two. If we ever add react-dom/server support to single-spa-react, then we might want separate bundles for browser/node.

> There is a naming convention with cjs, mjs and js. In order to use it as ESM, the file names for esm entries should end with .mjs or the package type should be "module", but then CommonJS entries should end with .cjs

I'd prefer setting `"type": "module"` and using `.js` for ESM code and `.cjs` for CJS code

> Then another question arises. Should we have copies of the same file with different names, f.e: ./lib/esm/single-spa-react.mjs and ./lib/esm/single-spa-react.js

I don't think that this has any benefit or is needed for ESM after we set the `type` to be `module`. Regarding `cjs` vs `js` on the UMD build, I think that setting the file extension to `cjs` fits the semantics and would work fine. single-spa-react is usually loaded via webpack/rollup, which do not follow the same rules as NodeJS but are okay with cjs files afaik. The one thing to check is if unpkg.com and jsdelivr.net send appropriate content-type headers when you load single-spa-react from CDN. We should see a `text/javascript` content-type from those CDNs.

Here are example files on those cdns for CJS files:

https://unpkg.com/systemjs@6.3.2/dist/system-node.cjs
https://cdn.jsdelivr.net/npm/systemjs@6.3.2/dist/system-node.cjs

Checking each of those URLs shows that unpkg sends `text/plain` and jsdelivr sends `application/octet-stream`. That could be a problem for those use single-spa-react from CDN, as browsers may not load those properly.

Perhaps this is reason to introduce the node/browser distinction (contradicting what I said earlier in this comment):

```json
{
"main": "lib/umd/single-spa-react.js",
"type": "module",
"exports": {
    ".": {
      "import": "./lib/esm/single-spa-react.js",
      "require": "./lib/cjs/single-spa-react.cjs"
    },
    "./parcel": {
      "import": "./lib/esm/parcel.js",
      "require": "./lib/cjs/parcel.cjs"
    }
  }
}
```

The esm directory would be usable in NodeJS, webpack/rollup, and browsers. The UMD directory would usable in browsers. And the CJS directory would be usable in old NodeJS?

Thoughts on this?

Reply:
I think given the problem with the content type, I think it is right idea to provide `cjs` for node and `umd` for browser. 

Also, if there are already capabilities to distinguish browser vs node in the "exports map", we may include it now:

```
  "main": "lib/umd/single-spa-react.js",
  "type": "module",
  "exports": {
    ".": {
      "import": "./lib/esm/single-spa-react.js",
      "require": {
        "browser": "./lib/umd/single-spa-react.js",
        "node": "./lib/cjs/single-spa-react.cjs"
      }
    },
    "./parcel": {
      "import": "./lib/esm/parcel.js",
      "require": {
        "browser": "./lib/umd/parcel.js",
        "node": "./lib/cjs/parcel.cjs"
      }
    }
  }
```

I also checked Node's compatibility of LTS versions with the changes above:
- v12 works same as v14;
- v10 imports the main entry as expected, but for the parcel it requires to provide the complete path ('single-spa-react/lib/umd/parcel' or 'single-spa-react/lib/cjs/parce.cjsl), but I think we cannot do anything about it.